File: Cryptography-and-Network-Security-Lab/Lab-02/transposition.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define MAX 500

// Create alphabetical order of key
void getOrder(char key[], int order[], int n)
{
    for (int i = 0; i < n; i++)
    {
        order[i] = 0;

        for (int j = 0; j < n; j++)
        {
            if (key[j] < key[i])
            {
                order[i]++;
            }
            else if (key[j] == key[i] && j < i)
            {
                order[i]++;
            }
        }
    }
}

// Encryption
void encrypt(char text[], char key[])
{
    int keyLen = strlen(key);
    int textLen = strlen(text);

    char clean[MAX];
    int n = 0;

    // Remove spaces and keep alphabets
    for (int i = 0; i < textLen; i++)
    {
        if (isalpha(text[i]))
        {
            clean[n++] = toupper(text[i]);
        }
    }

    clean[n] = '\0';

    // Calculate number of rows
    int rows = (n + keyLen - 1) / keyLen;

    // Create matrix
    char matrix[rows][keyLen];

    int index = 0;

    // Fill matrix row-wise
    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < keyLen; j++)
        {
            if (index < n)
            {
                matrix[i][j] = clean[index++];
            }
            else
            {
                matrix[i][j] = 'X';
            }
        }
    }

    // Find column order
    int order[keyLen];

    getOrder(key, order, keyLen);

    printf("\nMatrix:\n");

    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < keyLen; j++)
        {
            printf("%c ", matrix[i][j]);
        }
        printf("\n");
    }

    printf("\nEncrypted text: ");

    // Read columns according to key order
    for (int number = 0; number < keyLen; number++)
    {
        for (int col = 0; col < keyLen; col++)
        {
            if (order[col] == number)
            {
                for (int row = 0; row < rows; row++)
                {
                    printf("%c", matrix[row][col]);
                }

                break;
            }
        }
    }

    printf("\n");
}

// Decryption
void decrypt(char cipher[], char key[])
{
    int keyLen = strlen(key);
    int cipherLen = strlen(cipher);

    char clean[MAX];
    int n = 0;

    // Remove spaces
    for (int i = 0; i < cipherLen; i++)
    {
        if (isalpha(cipher[i]))
        {
            clean[n++] = toupper(cipher[i]);
        }
    }

    clean[n] = '\0';

    if (n % keyLen != 0)
    {
        printf("\nInvalid ciphertext length.\n");
        printf("Ciphertext length must be divisible by key length.\n");
        return;
    }

    int rows = n / keyLen;

    char matrix[rows][keyLen];

    int order[keyLen];

    getOrder(key, order, keyLen);

    int index = 0;

    // Fill columns according to key order
    for (int number = 0; number < keyLen; number++)
    {
        for (int col = 0; col < keyLen; col++)
        {
            if (order[col] == number)
            {
                for (int row = 0; row < rows; row++)
                {
                    matrix[row][col] = clean[index++];
                }

                break;
            }
        }
    }

    printf("\nDecrypted text: ");

    // Read row-wise
    for (int row = 0; row < rows; row++)
    {
        for (int col = 0; col < keyLen; col++)
        {
            printf("%c", matrix[row][col]);
        }
    }

    printf("\n");
}

int main()
{
    char key[MAX];
    char text[MAX];
    int choice;

    printf("===== COLUMNAR TRANSPOSITION CIPHER =====\n");

    printf("\nEnter key: ");
    scanf("%s", key);

    // Convert key to uppercase
    for (int i = 0; key[i] != '\0'; i++)
    {
        key[i] = toupper(key[i]);
    }

    printf("\n1. Encryption");
    printf("\n2. Decryption");

    printf("\n\nEnter your choice: ");
    scanf("%d", &choice);

    getchar();

    printf("Enter text: ");
    fgets(text, sizeof(text), stdin);

    text[strcspn(text, "\n")] = '\0';

    if (choice == 1)
    {
        encrypt(text, key);
    }
    else if (choice == 2)
    {
        decrypt(text, key);
    }
    else
    {
        printf("\nInvalid choice!\n");
    }

    return 0;
}

File: Cryptography-and-Network-Security-Lab/Lab-02/transposition.py
def get_order(key):
    # Sort according to alphabet.
    # enumerate() also handles repeated letters
    # from left to right.
    return sorted(range(len(key)), key=lambda i: (key[i], i))


def encrypt(text, key):

    # Remove spaces and keep alphabets
    text = ''.join(
        ch.upper()
        for ch in text
        if ch.isalpha()
    )

    key = key.upper()

    columns = len(key)

    # Add X padding
    while len(text) % columns != 0:
        text += 'X'

    rows = len(text) // columns

    # Create matrix
    matrix = []

    index = 0

    for i in range(rows):
        row = []

        for j in range(columns):
            row.append(text[index])
            index += 1

        matrix.append(row)

    print("\nMatrix:")

    for row in matrix:
        print(" ".join(row))

    # Determine column order
    order = get_order(key)

    print("\nColumn order:", end=" ")

    for col in order:
        print(col + 1, end=" ")

    print()

    # Read columns
    result = ""

    for col in order:
        for row in range(rows):
            result += matrix[row][col]

    return result


def decrypt(ciphertext, key):

    ciphertext = ''.join(
        ch.upper()
        for ch in ciphertext
        if ch.isalpha()
    )

    key = key.upper()

    columns = len(key)

    if len(ciphertext) % columns != 0:
        print("\nInvalid ciphertext length.")
        print("Ciphertext length must be divisible by key length.")
        return None

    rows = len(ciphertext) // columns

    # Empty matrix
    matrix = [
        [''] * columns
        for _ in range(rows)
    ]

    order = get_order(key)

    index = 0

    # Put ciphertext column-wise
    for col in order:

        for row in range(rows):

            matrix[row][col] = ciphertext[index]

            index += 1

    print("\nMatrix:")

    for row in matrix:
        print(" ".join(row))

    # Read row-wise
    result = ""

    for row in range(rows):

        for col in range(columns):

            result += matrix[row][col]

    return result


print("===== COLUMNAR TRANSPOSITION CIPHER =====")

key = input("\nEnter key: ")

print("\n1. Encryption")
print("2. Decryption")

choice = int(input("\nEnter your choice: "))

text = input("Enter text: ")


if choice == 1:

    cipher = encrypt(text, key)

    print("\nEncrypted text:", cipher)


elif choice == 2:

    plain = decrypt(text, key)

    if plain is not None:
        print("\nDecrypted text:", plain)


else:

    print("\nInvalid choice!")